cloud/kubernetes/fluentd-configmap.yml

```diff
@@ -0,0 +1,82 @@
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: fluentd-config
+  namespace: kube-system
+  labels:
+    addonmanager.kubernetes.io/mode: Reconcile
+data:
+  containers.input.conf: |-
+    <source>
+      @type tail
+      path /var/log/containers/*.log
+      pos_file /var/log/containers.log.pos
+      # Tags at this point are in the format of:
+      # reform.var.log.containers.<POD_NAME>_<NAMESPACE_NAME>_<CONTAINER_NAME>-<CONTAINER_ID>.log
+      tag reform.*
+      read_from_head true
+      format multi_format
+      <pattern>
+        format json
+        time_key time
+        time_format %Y-%m-%dT%H:%M:%S.%NZ
+      </pattern>
+      <pattern>
+        format /^(?<time>.+) (?<stream>stdout|stderr) [^ ]* (?<log>.*)$/
+        time_format %Y-%m-%dT%H:%M:%S.%N%:z
+      </pattern>
+    </source>
+
+    <filter reform.**>
+      @type parser
+      format /^(?<severity>\w)(?<time>\d{4} [^\s]*)\s+(?<pid>\d+)\s+(?<source>[^ \]]+)\] (?<log>.*)/
+      reserve_data true
+      suppress_parse_error_log true
+      emit_invalid_record_to_error false
+      key_name log
+    </filter>
+
+    <filter reform.**>
```

Comment:
Can you explain this filter stanza for me? If it drops all logs that don't match the provided format, then what's the point of the `<match reform.**>` stanza below? And why does this stanza match on just `reform.**` while the next one matches more specifically on `reform.var.log.containers.cockroachdb**`?

More generally, a comment on each stanza explaining its purpose would be great for future maintainability.

Reply:
The filters are not exclusively ran:
- so the first `<filter reform.**>` is the default case matching on all log records. This is needed for all other containers running that are not CockroachDB.
- the second `<filter reform.**>` runs on all log lines that match the regex inside of it. One can add extra patterns (say for custom applications there).

Each filter will republish the log record into fluentd queue so it can be picked up by other filters, hence why we change the tag name to `parsed` after we are done with it in the second filter.

Lastly, we make sure to only `Match` once on log records from the CockroachDB container so that we update the severity text from `E` to `error` etc (normalising the severity text format).